Issue #906

andreasabel opened the pull request:
[ fixed #906 #942 #2068 #3136 #3431 #4391 #4418 ]

Termination checker now tries to reduce calls away
using non-recursive clauses.

This fixes the problem that for dependent copattern definitions
the earlier, non-recursive clauses get into non-guarded positions
in later clauses via Agda's constraint solver, and there confuse
the termination checker.

andreasabel committed:
[ #906 opt ] use SmallSet to manage AllowedReductions

Small sets are implemented as fixed-length arrays of Booleans.
Could be compressed for very small sets to a Word64, with bit-hackery.

Since these are queried upon every unfolding in reduce, fast lookup
might shave off a few computation cycles.
But I don't expect a noticeable speedup. :)

andreasabel committed:
[ #906 ] add clauseRecursive field to Clause

Doesn't do anything yet.  Intended to distinguish clauses that contain
recursive calls from those that don't.  The latter could fire to get rid
of offending shallow recursive calls during termination checking.  Such
are common in dependent copattern matching (#906).

andreasabel committed:
[ #906 ] RecCheck: compute clauseRecursive; TermCheck: proceed by SCCs

RecCheck.recursive now remembers which clause is recursive and also
returns the SCCs in the given mutual block.

This means that the termination checker can proceed on a by-SCC basis.

andreasabel committed:
[ fixed #906 #942 #2068 #3136 #3431 #4391 #4418 ]

Termination checker now tries to reduce calls away
using non-recursive clauses.

This fixes the problem that for dependent copattern definitions
the earlier, non-recursive clauses get into non-guarded positions
in later clauses via Agda's constraint solver, and there confuse
the termination checker.

andreasabel committed:
[ CHANGELOG ] Updated 2 links to the user manual

andreasabel committed:
[ fixed #3081 (testcase) ] which is #906 --without-K

andreasabel committed:
[ #906 CHANGELOG ]

nad commented:
I'm generally skeptical towards reducing things, as this may make Agda slower. Do you first try without reducing anything? And do you store the unreduced or the reduced clauses?

andreasabel commented:
> I'm generally skeptical towards reducing things, as this may make Agda slower. 

Not sure what happens here, though.  Could be a bit slower, because we attempt a reduction for each recursive call we find, and reduction might duplicate terms.  However, reduction might also discard terms, making the call extraction faster.  Getting rid of recursive calls makes the call graph smaller, which means faster call graph completion. 

> Do you first try without reducing anything? 

Could be an idea, but the performance impacts are unclear either way.

To estimate the impact, we could have a counter in the Benchmarking structure that is increased each time a reduction fires in the termination checker.

> And do you store the unreduced or the reduced clauses?

The reduction only happens locally, in the termination checker, thus nothing changes w.r.t. what is stored in the signature.  However, it is possible that some termination errors are now printed differently, if non-terminating calls have been altered by some reduction on the outside.  An example of this is likely a contrived one.

Let me know if you find any performance regressions or incomprehensible termination errors.

andreasabel committed:
[ #906 ] Benchmark number of reductions in TermCheck

To count the number or reductions actually taking place in the
termination checker, run

  agda -v profile:7 -v term.reduce:5 FOO.agda

E.g.

  make AGDA_OPTS="-v term.reduce:5" std-lib-test

gives

  termination-checker-reduced-nonrecursive-call 40

So reduction is fired only 40 times for the whole standard library.
Performance impact should be minimal.

andreasabel commented:
Added commit to collect some statistics.
To count the number or reductions actually taking place in the termination checker, run
```
  agda -v profile:7 -v term.reduce:5 FOO.agda
```
E.g.
```
  make AGDA_OPTS="-v term.reduce:5" std-lib-test
```
gives
```
  termination-checker-reduced-nonrecursive-call 40
```
So reduction is fired only 40 times for the whole standard library.
Performance impact should be minimal.

@nad: This might relieve your worries...!?
